src/core/socialprovider.js

```diff
@@ -233,44 +234,46 @@ XMPPSocialProvider.prototype.sendMessage = function(to, msg, continuation) {
-  // Sending all messages as type 'normal' means we can't communicate across
-  // different client types, but sending all as type 'chat' means messages
-  // will be broadcast to all clients.
-  var messageType = (this.vCardStore.getClient(to).status === 'ONLINE') ?
-      'normal' : 'chat';
-  
-  try {
-    // After each message is received, reset the timeout to
-    // wait for at least 100ms to batch other messages received 
-    // in that window. However, if the oldest message in the batch 
-    // was received over 2s ago, don't reset the timeout, and 
-    // just allow the current timeout to execute.
-    this.messages.push(msg);
-    if (!this.sendMessagesTimeout) {
-      this.timeOfFirstMessageInBatch = Date.now();
-    }
-    if ((Date.now() - this.timeOfFirstMessageInBatch < 2000) ||
-        !this.sendMessagesTimeout) {
-      clearTimeout(this.sendMessagesTimeout);
-      this.sendMessagesTimeout = setTimeout(function() {
-        this.client.send(new window.XMPP.Element('message', {
-          to: to,
-          type: messageType
-        }).c('body').t(JSON.stringify(this.messages)));
-        this.messages = [];
-        this.sendMessagesTimeout = null;
-      }.bind(this), 100);  
-    }
-  } catch(e) {
-    this.logger.error(e.stack);
-    continuation(undefined, {
-      errcode: 'UNKNOWN',
-      message: e.message
-    });
-    return;
+  // After each message is received, reset the timeout to
+  // wait for at least 100ms to batch other messages received 
+  // in that window. However, if the oldest message in the batch 
+  // was received over 2s ago, don't reset the timeout, and 
+  // just allow the current timeout to execute.
+  if (!this.messages[to]) {
+    this.messages[to] = [];
+  }
+  this.messages[to].push(msg);
+  if (!this.sendMessagesTimeout) {
+    this.timeOfFirstMessageInBatch = Date.now();
+  }
+  if ((Date.now() - this.timeOfFirstMessageInBatch < 2000) ||
+      !this.sendMessagesTimeout) {
+    clearTimeout(this.sendMessagesTimeout);
+    this.sendMessagesTimeout = setTimeout(function () {
+      Object.keys(this.messages).forEach(function (to) {
+        // If the destination client is ONLINE (i.e. using the same type of
+        // client) send this message with type 'normal' so it only reaches
+        // that client, otherwise use type 'chat' to send to all clients.
+        // Sending all messages as type 'normal' means we can't communicate
+        // across different client types, but sending all as type 'chat'
+        // means messages will be broadcast to all clients.
+        var messageType =
+            (this.vCardStore.getClient(to).status === 'ONLINE') ?
+                'normal' : 'chat';
+
+        try {
+          this.client.send(new window.XMPP.Element('message', {
+            to: to,
+            type: messageType
+          }).c('body').t(JSON.stringify(this.messages[to])));
+        } catch(e) {
+          this.logger.warn(e.message);
```

Comment:
I think it should be in freedom-social-xmpp, because the batching logic
will probably differ in each social network - some may not need batching at
all, some might have very strict limits.

On Thu Jan 08 2015 at 2:39:32 PM Will <notifications@github.com> wrote:

> In src/core/socialprovider.js
> <https://github.com/freedomjs/freedom-social-xmpp/pull/80#discussion-diff-22675783>
> :
>
> > +        // client) send this message with type 'normal' so it only reaches
> > +        // that client, otherwise use type 'chat' to send to all clients.
> > +        // Sending all messages as type 'normal' means we can't communicate
> > +        // across different client types, but sending all as type 'chat'
> > +        // means messages will be broadcast to all clients.
> > +        var messageType =
> > +            (this.vCardStore.getClient(to).status === 'ONLINE') ?
> > +                'normal' : 'chat';
> > +
> > +        try {
> > +          this.client.send(new window.XMPP.Element('message', {
> > +            to: to,
> > +            type: messageType
> > +          }).c('body').t(JSON.stringify(this.messages[to])));
> > +        } catch(e) {
> > +          this.logger.warn(e.message);
>
> I guess the question I had when doing this is whether this is the right
> place to do the batching, or whether this is something that should be done
> more generally at the uProxy level, and the XMPP provider should
> immediately send the batch when it gets it. thoughts?
> … <#msg-f:1489760129250352005_>
> On Thu Jan 08 2015 at 11:35:51 AM dborkan <notifications@github.com>
> wrote: In src/core/socialprovider.js <
> https://github.com/freedomjs/freedom-social-xmpp/pull/80#discussion-diff-22675533>
> : > + // client) send this message with type 'normal' so it only reaches >
> + // that client, otherwise use type 'chat' to send to all clients. > + //
> Sending all messages as type 'normal' means we can't communicate > + //
> across different client types, but sending all as type 'chat' > + // means
> messages will be broadcast to all clients. > + var messageType = > +
> (this.vCardStore.getClient(to).status === 'ONLINE') ? > + 'normal' :
> 'chat'; > + > + try { > + this.client.send(new
> window.XMPP.Element('message', { > + to: to, > + type: messageType > +
> }).c('body').t(JSON.stringify(this.messages[to]))); > + } catch(e) { > +
> this.logger.warn(e.message); Right now sendMessage is always invoking
> continuation() (success) regardless of whether or not we catch an error
> from this.client.send. Do you think it would be better to only invoke
> continuation() if we don't catch any error from this.client.send? Or would
> this be bad because it could potentially delay continuation by up to 2.1
> seconds? Currently there is no way for callers of sendMessage to detect an
> error — Reply to this email directly or view it on GitHub <
> https://github.com/freedomjs/freedom-social-xmpp/pull/80/files#r22675533>
> .
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/freedomjs/freedom-social-xmpp/pull/80/files#r22675783>
> .
>